File: android/app/src/main/java/host/exp/exponent/ExponentManifest.java
// Copyright 2015-present 650 Industries. All rights reserved.

package host.exp.exponent;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Debug;
import android.util.Log;
import android.util.LruCache;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import javax.inject.Inject;
import javax.inject.Singleton;

import host.exp.exponent.analytics.Analytics;
import host.exp.exponent.analytics.EXL;
import host.exp.exponent.exceptions.ManifestException;
import host.exp.exponent.kernel.Crypto;
import host.exp.exponent.kernel.ExponentUrls;
import host.exp.exponent.network.ExponentHttpClient;
import host.exp.exponent.network.ExponentNetwork;
import host.exp.exponent.utils.ColorParser;
import okhttp3.Call;
import okhttp3.Request;
import okhttp3.Response;

@Singleton
public class ExponentManifest {

  public interface ManifestListener {
    void onCompleted(JSONObject manifest);
    void onError(Exception e);
    void onError(String e);
  }

  public interface BitmapListener {
    void onLoadBitmap(Bitmap bitmap);
  }

  private static final String TAG = ExponentManifest.class.getSimpleName();

  public static final String MANIFEST_ID_KEY = "id";
  public static final String MANIFEST_NAME_KEY = "name";
  public static final String MANIFEST_APP_KEY_KEY = "appKey";
  public static final String MANIFEST_SDK_VERSION_KEY = "sdkVersion";
  public static final String MANIFEST_IS_VERIFIED_KEY = "isVerified";
  public static final String MANIFEST_ICON_URL_KEY = "iconUrl";
  public static final String MANIFEST_PRIMARY_COLOR_KEY = "primaryColor";
  public static final String MANIFEST_ORIENTATION_KEY = "orientation";
  public static final String MANIFEST_DEVELOPER_KEY = "developer";
  public static final String MANIFEST_PACKAGER_OPTS_KEY = "packagerOpts";
  public static final String MANIFEST_PACKAGER_OPTS_DEV_KEY = "dev";
  public static final String MANIFEST_BUNDLE_URL_KEY = "bundleUrl";
  public static final String MANIFEST_STATUS_BAR_COLOR = "androidStatusBarColor";
  public static final String MANIFEST_HIDE_EXPONENT_NOTIFICATION_KEY = "androidHideExponentNotificationInShellApp";

  // Notification
  public static final String MANIFEST_NOTIFICATION_INFO_KEY = "notification";
  public static final String MANIFEST_NOTIFICATION_ICON_URL_KEY = "iconUrl";
  public static final String MANIFEST_NOTIFICATION_COLOR_KEY = "color";
  public static final String MANIFEST_NOTIFICATION_ANDROID_MODE = "androidMode";
  public static final String MANIFEST_NOTIFICATION_ANDROID_COLLAPSED_TITLE = "androidCollapsedTitle";

  // Debugging
  public static final String MANIFEST_DEBUGGER_HOST_KEY = "debuggerHost";
  public static final String MANIFEST_MAIN_MODULE_NAME_KEY = "mainModuleName";

  // Loading
  public static final String MANIFEST_LOADING_INFO_KEY = "loading";
  public static final String MANIFEST_LOADING_ICON_URL = "iconUrl";
  public static final String MANIFEST_LOADING_EXPONENT_ICON_COLOR = "exponentIconColor";
  public static final String MANIFEST_LOADING_EXPONENT_ICON_GRAYSCALE = "exponentIconGrayscale";
  public static final String MANIFEST_LOADING_BACKGROUND_IMAGE_URL = "backgroundImageUrl";
  public static final String MANIFEST_LOADING_BACKGROUND_COLOR = "backgroundColor";
  public static final String MANIFEST_LOADING_HIDE_EXPONENT_TEXT_KEY = "hideExponentText";

  private static final int MAX_BITMAP_SIZE = 192;
  private static final String REDIRECT_SNIPPET = "exp.host/--/to-exp/";

  Context mContext;
  ExponentNetwork mExponentNetwork;
  Crypto mCrypto;
  private LruCache<String, Bitmap> mMemoryCache;

  @Inject
  public ExponentManifest(Context context, ExponentNetwork exponentNetwork, Crypto crypto) {
    mContext = context;
    mExponentNetwork = exponentNetwork;
    mCrypto = crypto;

    int maxMemory = (int) (Runtime.getRuntime().maxMemory() / 1024);
    // Use 1/16th of the available memory for this memory cache.
    final int cacheSize = maxMemory / 16;
    mMemoryCache = new LruCache<String, Bitmap>(cacheSize) {
      @Override
      protected int sizeOf(String key, Bitmap bitmap) {
        return bitmap.getByteCount() / 1024;
      }
    };
  }

  public static boolean booleanValue(final JSONObject manifest, final String propertyKey, final boolean defaultValue) {
    String value = manifest.optString(propertyKey);
    if (value == null) {
      return defaultValue;
    } else if (value.toLowerCase().equals("true")) {
      return true;
    } else if (value.toLowerCase().equals("false")) {
      return false;
    } else {
      return defaultValue;
    }
  }

  public void fetchManifest(final String manifestUrl, final ManifestListener listener) {
    Analytics.markEvent(Analytics.TimedEvent.STARTED_FETCHING_MANIFEST);

    String realManifestUrl = manifestUrl;
    if (manifestUrl.contains(REDIRECT_SNIPPET)) {
      // Redirect urls look like "https://exp.host/--/to-exp/exp%3A%2F%2Fgj-5x6.jesse.internal.exp.direct%3A80".
      // Android is crazy and catches this url with this intent filter:
      //  <data
      //    android:host="*.exp.direct"
      //    android:pathPattern=".*"
      //    android:scheme="http"/>
      //  <data
      //    android:host="*.exp.direct"
      //    android:pathPattern=".*"
      //    android:scheme="https"/>
      // so we have to add some special logic to handle that. This is than handling arbitrary HTTP 301s and 302
      // because we need to add /index.exp to the paths.
      realManifestUrl = Uri.decode(realManifestUrl.substring(realManifestUrl.indexOf(REDIRECT_SNIPPET) + REDIRECT_SNIPPET.length()));
    }

    String httpManifestUrl = ExponentUrls.toHttp(realManifestUrl);
    final boolean isSecureDomain = httpManifestUrl.startsWith(Constants.API_HOST);

    // Append index.exp to path
    Uri uri = Uri.parse(httpManifestUrl);
    String newPath = uri.getPath();
    if (newPath == null) {
      newPath = "";
    }
    if (!newPath.endsWith("/")) {
      newPath += "/";
    }
    newPath += "index.exp";
    httpManifestUrl = uri.buildUpon().encodedPath(newPath).build().toString();

    // Fetch manifest
    final boolean shouldRequestSignedManifest = !isSecureDomain;
    Request.Builder requestBuilder = ExponentUrls.addExponentHeadersToUrl(httpManifestUrl);
    if (shouldRequestSignedManifest) {
      requestBuilder.header("Exponent-Accept-Signature", "true");
    }

    Analytics.markEvent(Analytics.TimedEvent.STARTED_MANIFEST_NETWORK_REQUEST);
    if (Constants.DEBUG_MANIFEST_METHOD_TRACING) {
      Debug.startMethodTracing("manifest");
    }
    mExponentNetwork.getClient().callSafe(requestBuilder.build(), new ExponentHttpClient.SafeCallback() {
      @Override
      public void onFailure(Call call, IOException e) {
        listener.onError(new ManifestException(e, manifestUrl));
      }

      @Override
      public void onResponse(Call call, Response response) {
        if (!response.isSuccessful()) {
          listener.onError(new ManifestException(null, manifestUrl));
          return;
        }

        if (Constants.DEBUG_MANIFEST_METHOD_TRACING) {
          Debug.stopMethodTracing();
        }
        Analytics.markEvent(Analytics.TimedEvent.FINISHED_MANIFEST_NETWORK_REQUEST);
        try {
          String manifestString = response.body().string();
          fetchManifestStep2(manifestString, isSecureDomain, shouldRequestSignedManifest, listener);
        } catch (JSONException e) {
          listener.onError(e);
        } catch (IOException e) {
          listener.onError(e);
        }
      }

      @Override
      public void onErrorCacheResponse(Call call, Response response) {
        EXL.d(TAG, "Initial HTTP request failed. Using cached or embedded response.");
        onResponse(call, response);
      }
    });
  }

  private void fetchManifestStep2(final String manifestString, final boolean isSecureDomain,
                                  final boolean shouldRequestSignedManifest, final ManifestListener listener) throws JSONException {
    final JSONObject manifest = new JSONObject(manifestString);
    final boolean isVerified = isSecureDomain;
    if (shouldRequestSignedManifest && manifest.has("manifestString") && manifest.has("signature")) {
      final JSONObject innerManifest = new JSONObject(manifest.getString("manifestString"));
      mCrypto.verifyPublicRSASignature(Constants.API_HOST + "/--/manifest-public-key",
          manifest.getString("manifestString"), manifest.getString("signature"), new Crypto.RSASignatureListener() {
        @Override
        public void onError(String errorMessage) {
          Log.w(TAG, errorMessage);
          fetchManifestStep3(innerManifest, isVerified, listener);
        }

        @Override
        public void onCompleted(boolean isValid) {
          fetchManifestStep3(innerManifest, isValid, listener);
        }
      });
    } else {
      fetchManifestStep3(manifest, isVerified, listener);
    }
  }

  private void fetchManifestStep3(final JSONObject manifest, final boolean isVerified, final ManifestListener listener) {
    try {
      manifest.put("isVerified", isVerified);
    } catch (JSONException e) {
      listener.onError(e);
    }

    if (!manifest.has("bundleUrl")) {
      listener.onError("No bundleUrl in manifest");
    }

    EXL.d(TAG, "Done fetching manifest");
    Analytics.markEvent(Analytics.TimedEvent.FINISHED_FETCHING_MANIFEST);
    listener.onCompleted(manifest);
  }

  public JSONObject normalizeManifest(final String manifestUrl, final JSONObject manifest) throws JSONException {
    if (!manifest.has(MANIFEST_ID_KEY)) {
      manifest.put(MANIFEST_ID_KEY, manifestUrl);
    }

    if (!manifest.has(MANIFEST_NAME_KEY)) {
      manifest.put(MANIFEST_NAME_KEY, "My New Experience");
    }

    if (!manifest.has(MANIFEST_PRIMARY_COLOR_KEY)) {
      manifest.put(MANIFEST_PRIMARY_COLOR_KEY, "#023C69");
    }

    if (!manifest.has(MANIFEST_ICON_URL_KEY)) {
      manifest.put(MANIFEST_ICON_URL_KEY, "https://d3lwq5rlu14cro.cloudfront.net/ExponentEmptyManifest_192.png");
    }

    if (!manifest.has(MANIFEST_ORIENTATION_KEY)) {
      manifest.put(MANIFEST_ORIENTATION_KEY, "default");
    }

    return manifest;
  }

  public void loadIconBitmap(final String iconUrl, final BitmapListener listener) {
    if (iconUrl != null && !iconUrl.isEmpty()) {
      Bitmap cachedBitmap = mMemoryCache.get(iconUrl);
      if (cachedBitmap != null) {
        listener.onLoadBitmap(cachedBitmap);
        return;
      }

      new AsyncTask<Void, Void, Bitmap>() {

        @Override
        protected Bitmap doInBackground(Void... params) {
          try {
            // TODO: inject shared OkHttp client
            URL url = new URL(iconUrl);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setDoInput(true);
            connection.connect();
            InputStream input = connection.getInputStream();

            Bitmap bitmap = BitmapFactory.decodeStream(input);
            int width = bitmap.getWidth();
            int height = bitmap.getHeight();
            if (width <= MAX_BITMAP_SIZE && height <= MAX_BITMAP_SIZE) {
              mMemoryCache.put(iconUrl, bitmap);
              return bitmap;
            }

            int maxDimension = Math.max(width, height);
            float scaledWidth = (((float) width) * MAX_BITMAP_SIZE) / maxDimension;
            float scaledHeight = (((float) height) * MAX_BITMAP_SIZE) / maxDimension;
            Bitmap scaledBitmap = Bitmap.createScaledBitmap(bitmap, (int) scaledWidth, (int) scaledHeight, true);
            mMemoryCache.put(iconUrl, scaledBitmap);
            return scaledBitmap;
          } catch (IOException e) {
            EXL.e(TAG, e);
            return BitmapFactory.decodeResource(mContext.getResources(), R.mipmap.ic_launcher);
          } catch (RuntimeException e) {
            EXL.e(TAG, e);
            return BitmapFactory.decodeResource(mContext.getResources(), R.mipmap.ic_launcher);
          }
        }

        @Override
        protected void onPostExecute(Bitmap result) {
          listener.onLoadBitmap(result);
        }
      }.execute();
    } else {
      Bitmap bitmap = BitmapFactory.decodeResource(mContext.getResources(), R.mipmap.ic_launcher);
      listener.onLoadBitmap(bitmap);
    }
  }

  public int getColorFromManifest(final JSONObject manifest) {
    String colorString = manifest.optString(MANIFEST_PRIMARY_COLOR_KEY);
    if (colorString != null && ColorParser.isValid(colorString)) {
      return Color.parseColor(colorString);
    } else {
      return R.color.colorPrimary;
    }
  }
}

File: apps/expo-go/android/expoview/src/main/java/host/exp/exponent/ExponentManifest.kt
// Copyright 2015-present 650 Industries. All rights reserved.
package host.exp.exponent

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.net.Uri
import android.os.AsyncTask
import android.text.TextUtils
import android.util.LruCache
import expo.modules.manifests.core.Manifest
import host.exp.exponent.analytics.EXL
import host.exp.exponent.generated.ExponentBuildConstants
import host.exp.exponent.kernel.ExponentUrls
import host.exp.exponent.kernel.KernelProvider
import host.exp.exponent.storage.ExponentSharedPreferences
import host.exp.exponent.utils.ColorParser
import host.exp.expoview.R
import org.apache.commons.io.IOUtils
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.max

data class ManifestAndAssetRequestHeaders(val manifest: Manifest, val assetRequestHeaders: JSONObject)

@Singleton
class ExponentManifest @Inject constructor(
  var context: Context,
  var exponentSharedPreferences: ExponentSharedPreferences
) {
  interface BitmapListener {
    fun onLoadBitmap(bitmap: Bitmap?)
  }

  private val memoryCache: LruCache<String, Bitmap>

  fun loadIconBitmap(iconUrl: String?, listener: BitmapListener) {
    val icon = getIconFromCache(iconUrl)
    if (icon != null) {
      listener.onLoadBitmap(icon)
      return
    }
    object : AsyncTask<Void?, Void?, Bitmap>() {
      override fun doInBackground(vararg p0: Void?): Bitmap? {
        return loadIconTask(iconUrl)
      }

      override fun onPostExecute(result: Bitmap?) {
        listener.onLoadBitmap(result)
      }
    }.execute()
  }

  private fun getIconFromCache(iconUrl: String?): Bitmap? {
    return if (iconUrl == null || TextUtils.isEmpty(iconUrl)) {
      BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)
    } else {
      memoryCache[iconUrl]
    }
  }

  private fun loadIconTask(iconUrl: String?): Bitmap? {
    return try {
      // TODO: inject shared OkHttp client
      val url = URL(iconUrl)
      val connection = url.openConnection() as HttpURLConnection
      connection.doInput = true
      connection.connect()
      val input = connection.inputStream
      val bitmap = BitmapFactory.decodeStream(input)
      val width = bitmap.width
      val height = bitmap.height
      if (width <= MAX_BITMAP_SIZE && height <= MAX_BITMAP_SIZE) {
        memoryCache.put(iconUrl, bitmap)
        return bitmap
      }
      val maxDimension = max(width, height)
      val scaledWidth = width.toFloat() * MAX_BITMAP_SIZE / maxDimension
      val scaledHeight = height.toFloat() * MAX_BITMAP_SIZE / maxDimension
      val scaledBitmap =
        Bitmap.createScaledBitmap(bitmap, scaledWidth.toInt(), scaledHeight.toInt(), true)
      memoryCache.put(iconUrl, scaledBitmap)
      scaledBitmap
    } catch (e: IOException) {
      EXL.e(TAG, e)
      BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)
    } catch (e: Throwable) {
      EXL.e(TAG, e)
      BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)
    }
  }

  fun getColorFromManifest(manifest: Manifest): Int {
    val colorString = manifest.getPrimaryColor()
    return if (colorString != null && ColorParser.isValid(colorString)) {
      Color.parseColor(colorString)
    } else {
      R.color.colorPrimary
    }
  }

  private fun getLocalKernelManifestAndAssetRequestHeaders(): ManifestAndAssetRequestHeaders = try {
    val manifestAndAssetRequestHeaders = JSONObject(ExponentBuildConstants.getBuildMachineKernelManifestAndAssetRequestHeaders())
    val manifest = manifestAndAssetRequestHeaders.getJSONObject("manifest")
    val assetRequestHeaders = manifestAndAssetRequestHeaders.getJSONObject("assetRequestHeaders")
    manifest.put(MANIFEST_IS_VERIFIED_KEY, true)
    ManifestAndAssetRequestHeaders(Manifest.fromManifestJson(manifest), assetRequestHeaders)
  } catch (e: JSONException) {
    throw RuntimeException("Can't get local manifest: $e")
  }

  private fun getEmbeddedKernelManifest(): Manifest? = try {
    val inputStream = context.assets.open(EMBEDDED_KERNEL_MANIFEST_ASSET)
    val jsonString = IOUtils.toString(inputStream)
    val manifest = JSONObject(jsonString)
    manifest.put(MANIFEST_IS_VERIFIED_KEY, true)
    Manifest.fromManifestJson(manifest)
  } catch (e: Exception) {
    KernelProvider.instance.handleError(e)
    null
  }

  fun getKernelManifestAndAssetRequestHeaders(): ManifestAndAssetRequestHeaders {
    val manifestAndAssetRequestHeaders: ManifestAndAssetRequestHeaders
    val log: String
    if (exponentSharedPreferences.shouldUseEmbeddedKernel()) {
      log = "Using embedded Expo kernel manifest"
      manifestAndAssetRequestHeaders = ManifestAndAssetRequestHeaders(getEmbeddedKernelManifest()!!, JSONObject())
    } else {
      log = "Using local Expo kernel manifest"
      manifestAndAssetRequestHeaders = getLocalKernelManifestAndAssetRequestHeaders()
    }
    if (!hasShownKernelManifestLog) {
      hasShownKernelManifestLog = true
      EXL.d(TAG, log + ": " + manifestAndAssetRequestHeaders.manifest.toString())
    }
    return manifestAndAssetRequestHeaders
  }

  companion object {
    private val TAG = ExponentManifest::class.java.simpleName

    const val MANIFEST_STRING_KEY = "manifestString"
    const val MANIFEST_SIGNATURE_KEY = "signature"
    const val MANIFEST_ID_KEY = "id"
    const val MANIFEST_NAME_KEY = "name"
    const val MANIFEST_APP_KEY_KEY = "appKey"
    const val MANIFEST_SDK_VERSION_KEY = "sdkVersion"
    const val MANIFEST_IS_VERIFIED_KEY = "isVerified"
    const val MANIFEST_ICON_URL_KEY = "iconUrl"
    const val MANIFEST_BACKGROUND_COLOR_KEY = "backgroundColor"
    const val MANIFEST_PRIMARY_COLOR_KEY = "primaryColor"
    const val MANIFEST_ORIENTATION_KEY = "orientation"
    const val MANIFEST_DEVELOPER_KEY = "developer"
    const val MANIFEST_DEVELOPER_TOOL_KEY = "tool"
    const val MANIFEST_PACKAGER_OPTS_KEY = "packagerOpts"
    const val MANIFEST_PACKAGER_OPTS_DEV_KEY = "dev"
    const val MANIFEST_BUNDLE_URL_KEY = "bundleUrl"
    const val MANIFEST_REVISION_ID_KEY = "revisionId"
    const val MANIFEST_PUBLISHED_TIME_KEY = "publishedTime"
    const val MANIFEST_COMMIT_TIME_KEY = "commitTime"
    const val MANIFEST_LOADED_FROM_CACHE_KEY = "loadedFromCache"
    const val MANIFEST_SLUG = "slug"
    const val MANIFEST_ANDROID_INFO_KEY = "android"
    const val MANIFEST_KEYBOARD_LAYOUT_MODE_KEY = "softwareKeyboardLayoutMode"

    // Statusbar
    const val MANIFEST_STATUS_BAR_KEY = "androidStatusBar"
    const val MANIFEST_STATUS_BAR_APPEARANCE = "barStyle"
    const val MANIFEST_STATUS_BAR_BACKGROUND_COLOR = "backgroundColor"
    const val MANIFEST_STATUS_BAR_HIDDEN = "hidden"
    const val MANIFEST_STATUS_BAR_TRANSLUCENT = "translucent"

    // NavigationBar
    const val MANIFEST_NAVIGATION_BAR_KEY = "androidNavigationBar"
    const val MANIFEST_NAVIGATION_BAR_VISIBLILITY = "visible"
    const val MANIFEST_NAVIGATION_BAR_APPEARANCE = "barStyle"
    const val MANIFEST_NAVIGATION_BAR_BACKGROUND_COLOR = "backgroundColor"

    // Notification
    const val MANIFEST_NOTIFICATION_INFO_KEY = "notification"
    const val MANIFEST_NOTIFICATION_ICON_URL_KEY = "iconUrl"
    const val MANIFEST_NOTIFICATION_COLOR_KEY = "color"
    const val MANIFEST_NOTIFICATION_ANDROID_MODE = "androidMode"
    const val MANIFEST_NOTIFICATION_ANDROID_COLLAPSED_TITLE = "androidCollapsedTitle"

    // Debugging
    const val MANIFEST_DEBUGGER_HOST_KEY = "debuggerHost"
    const val MANIFEST_MAIN_MODULE_NAME_KEY = "mainModuleName"

    // Splash
    const val MANIFEST_SPLASH_INFO_KEY = "splash"
    const val MANIFEST_SPLASH_IMAGE_URL_KEY = "imageUrl"
    const val MANIFEST_SPLASH_RESIZE_MODE_KEY = "resizeMode"
    const val MANIFEST_SPLASH_BACKGROUND_COLOR_KEY = "backgroundColor"

    // Updates
    const val MANIFEST_UPDATES_INFO_KEY = "updates"
    const val MANIFEST_UPDATES_TIMEOUT_KEY = "fallbackToCacheTimeout"
    const val MANIFEST_UPDATES_CHECK_AUTOMATICALLY_KEY = "checkAutomatically"
    const val MANIFEST_UPDATES_CHECK_AUTOMATICALLY_ON_LOAD = "ON_LOAD"
    const val MANIFEST_UPDATES_CHECK_AUTOMATICALLY_ON_ERROR = "ON_ERROR_RECOVERY"

    // Development client
    const val MANIFEST_DEVELOPMENT_CLIENT_KEY = "developmentClient"
    const val MANIFEST_DEVELOPMENT_CLIENT_SILENT_LAUNCH_KEY = "silentLaunch"
    const val DEEP_LINK_SEPARATOR = "--"
    const val DEEP_LINK_SEPARATOR_WITH_SLASH = "--/"
    const val QUERY_PARAM_KEY_RELEASE_CHANNEL = "release-channel"
    const val QUERY_PARAM_KEY_EXPO_UPDATES_RUNTIME_VERSION = "runtime-version"
    const val QUERY_PARAM_KEY_EXPO_UPDATES_CHANNEL_NAME = "channel-name"

    // snack
    const val QUERY_PARAM_KEY_SNACK_CHANNEL = "snack-channel"

    private const val MAX_BITMAP_SIZE = 192
    private const val REDIRECT_SNIPPET = "exp.host/--/to-exp/"
    private const val ANONYMOUS_SCOPE_KEY_PREFIX = "@anonymous/"
    private const val EMBEDDED_KERNEL_MANIFEST_ASSET = "kernel-manifest.json"
    private const val EXPONENT_SERVER_HEADER = "Exponent-Server"

    private var hasShownKernelManifestLog = false

    fun isAnonymousExperience(manifest: Manifest): Boolean {
      return try {
        manifest.getScopeKey().startsWith(ANONYMOUS_SCOPE_KEY_PREFIX)
      } catch (e: JSONException) {
        false
      }
    }

    fun httpManifestUrl(manifestUrl: String): Uri {
      return httpManifestUrlBuilder(manifestUrl).build()
    }

    private fun httpManifestUrlBuilder(manifestUrl: String): Uri.Builder {
      var realManifestUrl = manifestUrl
      if (manifestUrl.contains(REDIRECT_SNIPPET)) {
        // Redirect urls look like "https://exp.host/--/to-exp/exp%3A%2F%2Fgj-5x6.jesse.internal.exp.direct%3A80".
        // Android is crazy and catches this url with this intent filter:
        //  <data
        //    android:host="*.exp.direct"
        //    android:pathPattern=".*"
        //    android:scheme="http"/>
        //  <data
        //    android:host="*.exp.direct"
        //    android:pathPattern=".*"
        //    android:scheme="https"/>
        // so we have to add some special logic to handle that. This is than handling arbitrary HTTP 301s and 302
        realManifestUrl = Uri.decode(
          realManifestUrl.substring(
            realManifestUrl.indexOf(
              REDIRECT_SNIPPET
            ) + REDIRECT_SNIPPET.length
          )
        )
      }
      val httpManifestUrl = ExponentUrls.toHttp(realManifestUrl)
      val uri = Uri.parse(httpManifestUrl)
      var newPath = uri.path
      if (newPath == null) {
        newPath = ""
      }
      val deepLinkIndex = newPath.indexOf(DEEP_LINK_SEPARATOR_WITH_SLASH)
      if (deepLinkIndex > -1) {
        newPath = newPath.substring(0, deepLinkIndex)
      }
      return uri.buildUpon().encodedPath(newPath)
    }
  }

  init {
    val maxMemory = (Runtime.getRuntime().maxMemory() / 1024).toInt()
    // Use 1/16th of the available memory for this memory cache.
    val cacheSize = maxMemory / 16
    memoryCache = object : LruCache<String, Bitmap>(cacheSize) {
      override fun sizeOf(key: String?, bitmap: Bitmap): Int {
        return bitmap.byteCount / 1024
      }
    }
  }
}
